latent_programmer/decomposition_transformer_attention/decomposition_models.py

```diff
@@ -90,11 +92,45 @@ class DecomposeAttentionTransformer(nn.Module):
   def setup(self):
     base_config = self.config.base_config
 
-    self.encoder = base_models.TransformerIOEncoder(config=base_config,
-                                                    name='encoder')
-    # Shifting is done before call to decoder in order to compute masks.
+    is_initialized = (self.has_variable('relative_attention', 'max_input_distance') and
+                      self.has_variable('relative_attention', 'max_output_distance') and
+                      self.has_variable('relative_attention', 'max_io_distance') and
+                      self.has_variable('relative_attention', 'max_program_distance')
+                      )
+    if is_initialized and self.config.clip_relative_attention:
+      max_input_distance = self.max_input_distance.item()
+      max_output_distance = self.max_output_distance.item()
+      max_io_distance = self.max_io_distance.item()
+      max_program_distance = self.max_program_distance.item()
+
+      base_encoder_config = base_config.replace(
+          max_distance=max_input_distance,
+          num_relative_position_buckets=2 * max_input_distance,
+          max_output_distance=max_output_distance,
+          num_output_relative_position_buckets=2 * max_output_distance)
+      base_decoder_config=base_config.replace(
+          max_distance=max_io_distance,
+          num_relative_position_buckets=2 * max_io_distance,
+          max_output_distance=max_program_distance,
+          num_output_relative_position_buckets=2 * max_program_distance,
+          shift=False),  # Shifting is done separately in decoder.
+    else:
+      base_encoder_config = base_config
+      base_decoder_config = base_config.replace(shift=False)
+
+    self.encoder = base_models.TransformerIOEncoder(
+        config=base_encoder_config, name='encoder')
     self.decoder = base_models.TransformerDecoder(
-        config=base_config.replace(shift=False), name='decoder')
+        config=base_decoder_config, name='decoder')
+
+    self.max_input_distance = self.variable(
```

Comment:
can we put this inside an `if not is_initialized:` block? we don't want to overwrite them if they're already initialized